### DISCUSSION TITLE
Determine if job failed in wrapper action

### DISCUSSION THREAD

**Discussion by yvrhdn** [community/community#26451] [Actions]
<p>Hi, I’ve a question about the undocumented <strong>wrapper action</strong>.</p>
<p>The ‘post’ and ‘post-if’ directives are currently still undocumented, but you can already see this in action in <a href="https://github.com/actions/checkout/blob/master/action.yml#L74" rel="nofollow noopener">actions/checkout</a> and <a href="https://github.com/actions/cache/blob/master/action.yml#L20" rel="nofollow noopener">actions/cache</a>.</p>
<p>It is also mentioned here: <a href="https://github.com/actions/checkout/blob/master/action.yml#L74" rel="nofollow noopener">https://github.com/actions/checkout/blob/master/action.yml#L74</a></p>
<p>I’ve been able to get this working in my own action <a href="https://github.com/kvrhdn/gha-buildevents" rel="nofollow noopener">gha-buildevents</a>, but I’d like to be able to know whether the job failed or not.</p>
<p>Is it possible to determine the state of the job at this point? I couldn’t find any useful variables in the environment.</p>
<p>Thanks!</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/kvrhdn">@kvrhdn</a> ,</p>
<p>The wrapper action(post &amp; post-if) is still not documented. The post job step will be added with condition set to always() by default and can be overridden via post-if. Some samples below:</p>
<p><strong>Node.js action action.yaml changes</strong></p>
<p>Extend the runs section in the action.yml to include the script to run during cleanup.</p>
<pre><code>runs:
  using: 'node12'
  main: 'main.js'
  post: 'cleanup.js'
  post-if: 'failure()' // Optional
</code></pre>
<p><strong>Container action action.yaml changes</strong></p>
<p>Extend the runs section in the action.yml to include the script to run during cleanup.</p>
<pre><code>runs:
  using: 'docker'
  image: 'Dockerfile'
  entrypoint: '/main.sh'
  post-entrypoint: '/cleanup.sh'
  post-if: 'runner.os == linux' // Optional
</code></pre>
<p>Hope it helps!</p>

**Comment by yvrhdn** [comment]
<p>Hi <a class="mention" href="/weide-zhou">@weide-zhou</a>, thanks for the  quick reply!</p>
<p>I’ve managed to get the post job step working. But I would like to do something else within the step depending on the status of the build.</p>
<p>For example: the post job step should always run , but if:</p>
<ul>
<li>status == failure: send a trace telling the build failed</li>
<li>status == success: send a trace telling the build succeeded</li>
</ul>
<p>Is there some environment variable that contains the status of the job?</p>
<p>Or would it be possible to run a different script depending on the status? Something like:</p>
<pre><code>runs:
  using: 'node12'
  main: 'main.js'
  post-success: 'post-success.js'
  post-failure: 'post-failure.js'
</code></pre>
<p>My current workaround is to set an environment variables BUILD_SUCCESS as final step. So this variable will only be set if <em>all</em> steps executed sucessfully.</p>
<p>Example: <a href="https://github.com/kvrhdn/gha-buildevents/blob/master/.github/workflows/integration.yaml#L37" rel="nofollow noopener">https://github.com/kvrhdn/gha-buildevents/blob/master/.github/workflows/integration.yaml#L37</a></p>
<p>Cheers!</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/kvrhdn">@kvrhdn</a> ,</p>
<p>You can get the job status from github context : ${{ job.status }}, the value could be success, failure…etc.</p>
<p>Code sample as below:</p>
<pre><code>- name: job status
        if: always() # use if expression to make sure step is executed even job failed.
        run: echo ${{ job.status }}
</code></pre>
<p>‘post-success’&amp;‘post-failure’ is not accepted in action.yaml.</p>
<p><img src="https://user-images.githubusercontent.com/52772761/181097869-341a253f-fb06-47e6-b6e4-4fae5d2acfe8.png" alt="jobstatus.png" data-base62-sha1="2QZOhA2D10SzqcZB3hjcHgCaxBi" width="386" height="183"></p>
<p>Thanks.</p>

**Comment by yvrhdn** [comment]
<p>Thanks! I didn’t know about the job context. I’ve found my way to the documentation now as well.</p>
<p>Is it possible to access this variable from inside an action? I’m using the <a href="https://github.com/actions/toolkit" rel="nofollow noopener">actions/toolkit</a> but besides getting inputs and environment variables, it doesn’t expose any job infromation.</p>
<p>Would it be my best to fetch the status using the GitHub API? <a href="https://octokit.github.io/rest.js/v17#actions-get-workflow-run" rel="nofollow noopener">https://octokit.github.io/rest.js/v17#actions-get-workflow-run</a></p>

**Comment by yvrhdn** [comment]
<p>Thanks! I didn’t know about the job context. I’ve now also found the documentation for that :slight_smile:</p>
<p>Is it possible to get this value from within a Javascript action? I’m using actions/core from the toolkit, but this only contains functions to get inputs and environment variables. And as far as I can see, ‘job.context’ isn’t present in the environment variables.</p>
<p>I’ve tried getting the workflow status through the GitHub API, but this always returns ‘in_progress’.</p>
<p>Thanks!</p>

**Comment by yvrhdn** [comment, marked as answer]
<p>I’ve found a solution!</p>
<p>Inputs to the action are re-evaluated when passed to the post section. So if you have an input ‘job.status’, this is effectively updated with the actual job status at the end of the regular steps.</p>
<p>I made the action so you have to provide ‘job.status’ as input:</p>
<pre><code>- uses: kvrhdn/gha-buildevents@master
  with:
    job-status: ${{ job.status }}
    ...
</code></pre>
<p>If you enable debug logging, you can see that the post section re-evaluates all inputs!</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/030d105c87a8ed09519f53d190f657b57b327c8a.png" data-download-href="/uploads/short-url/qZq948jltn5Feip7oDNRZ7VDlg.png?dl=1" title="Screenshot 2020-04-29 at 03.00.14.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/7748404/181097875-c1a76fc4-171d-4def-933f-6ffc761e56bb.png" alt="Screenshot 2020-04-29 at 03.00.14.png" data-base62-sha1="qZq948jltn5Feip7oDNRZ7VDlg" width="690" height="370" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/0/030d105c87a8ed09519f53d190f657b57b327c8a_2_690x370.png, https://aws1.discourse-cdn.com/github/optimized/2X/0/030d105c87a8ed09519f53d190f657b57b327c8a_2_1035x555.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/0/030d105c87a8ed09519f53d190f657b57b327c8a_2_1380x740.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/030d105c87a8ed09519f53d190f657b57b327c8a_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot 2020-04-29 at 03.00.14.png</span><span class="informations">2050×1102 305 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Reply by Janno** [reply]
Thank you so much for documenting your findings. I was going insane trying to find a way to get the job status in a post action. I would have never found this solution without you!
